### PR TITLE
update toFieldPath policies with options compatible to upstream mergeOptions

### PR DESCRIPTION
**Pull request by ravilr**
- Updates https://github.com/crossplane-contrib/function-patch-and-transform/pull/91
- Partially addresses https://github.com/crossplane-contrib/function-patch-and-transform/issues/101 with native P&T compatible mergeOptions as per the proposal in https://github.com/crossplane-contrib/function-patch-and-transform/issues/101#issuecomment-2024155035

### CODE REVIEW

**Comment by ravilr** [comment]
@phisco @negz PTAL 🙏

File: input/v1beta1/resources_patches.go [hunk 1]
```diff
@@ -37,9 +37,13 @@ type ToFieldPathPolicy string
...
-	ToFieldPathPolicyReplace     ToFieldPathPolicy = "Replace"
-	ToFieldPathPolicyMergeObject ToFieldPathPolicy = "MergeObject"
-	ToFieldPathPolicyAppendArray ToFieldPathPolicy = "AppendArray"
+	ToFieldPathPolicyReplace                       ToFieldPathPolicy = "Replace"
+	ToFieldPathPolicyMergeObject                   ToFieldPathPolicy = "MergeObject" // Deprecated in favor of MergeObjects
+	ToFieldPathPolicyAppendArray                   ToFieldPathPolicy = "AppendArray" // Deprecated in favor of ForceMergeObjectsAppendArrays
+	ToFieldPathPolicyMergeObjects                  ToFieldPathPolicy = "MergeObjects"
+	ToFieldPathPolicyForceMergeObjectsAppendArrays ToFieldPathPolicy = "ForceMergeObjectsAppendArrays"
+	ToFieldPathPolicyForceMergeObjects             ToFieldPathPolicy = "ForceMergeObjects"
+	ToFieldPathPolicyMergeObjectsAppendArrays      ToFieldPathPolicy = "MergeObjectsAppendArrays"
```
Comment (negz) [inline input/v1beta1/resources_patches.go:46]:
```suggestion
	ToFieldPathPolicyReplace                       ToFieldPathPolicy = "Replace"
	ToFieldPathPolicyMergeObjects                  ToFieldPathPolicy = "MergeObjects"
	ToFieldPathPolicyMergeObjectsAppendArrays      ToFieldPathPolicy = "MergeObjectsAppendArrays"
	ToFieldPathPolicyForceMergeObjectsAppendArrays ToFieldPathPolicy = "ForceMergeObjectsAppendArrays"
	ToFieldPathPolicyForceMergeObjects             ToFieldPathPolicy = "ForceMergeObjects"


	ToFieldPathPolicyMergeObject                   ToFieldPathPolicy = "MergeObject" // Deprecated: Use MergeObjects, which is functionally identical.
	ToFieldPathPolicyAppendArray                   ToFieldPathPolicy = "AppendArray" // Deprecated: Use ForceMergeObjectsAppendArrays, which is functionally identical.	
```

Nit: Sorts these as replace, regular merge, force merge, deprecated stuff.

Also uses the https://go.dev/wiki/Deprecated comment notation. Though I'm not 100% sure that works as a trailing comment.

File: input/v1beta1/resources_patches.go [hunk 2]
```diff
@@ -54,10 +58,17 @@ type PatchPolicy struct {
...
+	// object with the target object, while keeping target object keys, but overwriting any array values or use
+	// 'ForceMergeObjectsAppendArrays' to recursively merge the patch object with the target object,
+	// overwriting target object keys, and appending any array values to target array values, or use
+	// 'ForceMergeObjects' to recursively merge the patch object with the target object, overwriting
+	// any target object keys, including array values, or use
+	// 'MergeObjectsAppendArrays' to recursively merge the patch object with the target object, while keeping
+	// target object keys and appending any array values to target array values.
+	// 'AppendArray' is deprecated, use 'ForceMergeObjectsAppendArrays' instead.
+	// 'MergeObject' is deprecated, use 'MergeObjects' instead.
+	// +kubebuilder:validation:Enum=Replace;MergeObject;AppendArray;MergeObjects;ForceMergeObjectsAppendArrays;ForceMergeObjects;MergeObjectsAppendArrays
```
Comment (negz) [inline input/v1beta1/resources_patches.go:71]: Nit: Maybe sort these as replace, regular merge, force merge, deprecated as well?

File: patches.go [hunk 3]
```diff
@@ -102,10 +102,14 @@ func toMergeOption(p PatchInterface) (mo *xpv1.MergeOptions, err error) {
...
-	case v1beta1.ToFieldPathPolicyAppendArray:
+	case v1beta1.ToFieldPathPolicyAppendArray, v1beta1.ToFieldPathPolicyForceMergeObjectsAppendArrays:
 		mo = &xpv1.MergeOptions{AppendSlice: ptr.To(true)}
-	case v1beta1.ToFieldPathPolicyMergeObject:
+	case v1beta1.ToFieldPathPolicyMergeObject, v1beta1.ToFieldPathPolicyMergeObjects:
 		mo = &xpv1.MergeOptions{KeepMapValues: ptr.To(true)}
+	case v1beta1.ToFieldPathPolicyForceMergeObjects:
+		mo = &xpv1.MergeOptions{KeepMapValues: ptr.To(false)}
+	case v1beta1.ToFieldPathPolicyMergeObjectsAppendArrays:
+		mo = &xpv1.MergeOptions{KeepMapValues: ptr.To(true), AppendSlice: ptr.To(true)}
```
Comment (negz) [inline patches.go:112]: I'd find this a little easier to parse if it was sorted:

1. Replace
2. MergeObjects, MergeObject
3. MergeObjectsAppendArrays
4. ForceMergeObjects
5. ForceMergeObjectsAppendArrays, AppendArray

**Review by negz — COMMENTED** [review]
Looks good to me, thank you!

My only comments are about sorting the list of policies.

File: input/v1beta1/resources_patches.go (hunk 1 above)
Comment (ravilr) [reply input/v1beta1/resources_patches.go]: Thanks, addressed.

File: input/v1beta1/resources_patches.go (hunk 2 above)
Comment (ravilr) [reply input/v1beta1/resources_patches.go]: Thanks, addressed.

File: patches.go (hunk 3 above)
Comment (ravilr) [reply patches.go]: Thanks, addressed.

**Review by negz — APPROVED** [review]

**Comment by negz** [comment]
@ravilr Aha, looks like the linter caught a few uses of the newly deprecated types in the codebase. Could you update those too?

**Review by phisco — APPROVED** [review]
Awesome! Thanks @ravilr!

**Pull request merged**